Orb tools 11 (#6560)
circleci/circleci-docs: jekyll/_cci2/testing-orbs.md

## Introduction
{: #introduction }

> Orbs are open source and contributions are expected, therefore, as with all software, setting up a solid testing pipeline for your orb is important. Because orbs are created in YAML, it may seem difficult to test them effectively. With the Orb Development Kit, there is a simple path to implementing a full range of robust tests for your orb.

Author: Ya actually the whole paragraph sounds weird. Going to replace this.

> Orbs are a critical component of a pipeline on CircleCI, responsible for installing tooling, executing tests, deploying applications, and more. So like any software, it is important to implement tests to protect the orb from breaking with new changes. Because orbs are developed in YAML, the testing process is a little different than for a programming language, but with the Orb Development Kit, there is a simple path to implementing a full range of robust tests for your orb.

Reviewer: ```suggestion
Orbs are a critical component of a pipeline on CircleCI, responsible for installing tooling, executing tests, deploying applications, and more. As with any software, it is important to implement tests to protect the orb from breaking with new changes. Because orbs are developed in YAML, the testing process is a little different than for a programming language. With the Orb Development Kit, there is a simple path to implementing a full range of robust tests for your orb.
```

#### Local Review
{: #local-review }

The "review" job is built using the [BATS-Core](https://github.com/bats-core/bats-core) bash automation testing framework. If you have bats installed locally, you can review your orb by running:

```shell
bats src/scripts/review.bats
```

Or, using CircleCI's Local Execute:
```shell
circleci local execute --job orb-tools/review
```

## Unit testing
{: #unit-testing }

If you are taking advantage of the Orb Development Kit's [`<<include(file)>>` file inclusion]({{site.baseurl}}/2.0/orb-concepts/#file-include-syntax) feature and `src/scripts` directory to store and source your bash files, you can write true integration tests for your scripts.

![Unit testing BASH with BATS-Core]({{site.baseurl}}/assets/img/docs/bats_tests_example.png)

If you have an orb with sufficiently complex internal scripts, you may want to implement unit tests for better code quality and easier local development testing.

For Bash unit testing, we recommend the [BATS-Core](https://github.com/bats-core/bats-core) library, which is an open source testing framework for Bash, analogous to [Jest](https://jestjs.io/) for JavaScript.

> CircleCI has created a [BATS orb](https://circleci.com/developer/orbs/orb/circleci/bats-core) to easily integrate BATS into your CircleCI Pipelines.

Reviewer 2: ```suggestion
CircleCI has created a [BATS orb](https://circleci.com/developer/orbs/orb/circleci/bats-core) to easily integrate BATS into your CircleCI pipelines.
```

To add BATS to your orb, follow these steps:

  1. Add a `tests` directory to your orb's `src` directory.
  2. Create your tests in the `tests` directory.
  3. Add the [circleci/bats](https://circleci.com/developer/orbs/orb/circleci/bats#usage-run-bats-tests) orb to your `config.yml` file.
  4. Add the `bats/run` job to the pre-publishing jobs in the `config.yml` file.

```
workflows:
  lint-pack:
    jobs:
      - orb-tools/lint:
          filters: *filters
      - orb-tools/pack:
          filters: *filters
      - orb-tools/review:
          filters: *filters
# Add bats
      - bats/run:
          filters: *filters
          path: ./src/tests
# ...
# And ensure to mark it as required in the publish job.
 - orb-tools/publish:
          requires:
            [orb-tools/lint, orb-tools/review, orb-tools/pack, shellcheck/check, bats/run]
```

Want to see how how CircleCI writes unit tests for Bash? Check out our [Slack orb](https://github.com/CircleCI-Public/slack-orb/blob/master/src/tests/notify.bats).

## Integration testing
{: #integration-testing }

> After validating, linting, shellchecking, and any other testing that you can perform on the source code is complete, we must test our orb's functionality in a real CircleCI config. In the second config file (`test-deploy.yml`), we will access the development version of the orb we published in the first config, and attempt to execute our orbs commands and jobs.

Reviewer 2: ```suggestion
After validating, linting, shellchecking, and any other testing that you can perform on the source code is complete, you must test your orb's functionality in a real CircleCI config. In the second config file (`test-deploy.yml`), you can access the development version of the orb you published in the first config, and attempt to execute your orbs commands and jobs.
```

### Testing orb commands
{: #testing-orb-commands }

> By default, when you author a new orb, you will have an example orb source which comes with a "greet" command. We want to test the greet command (and maybe other commands) in our `test-deploy` workflow as an integration test. We'll be able to execute the commands to validate they run without error, and we could even verify their functionality by running additional checks.

Reviewer 2: ```suggestion
By default, when you author a new orb, you will have an example orb source which comes with a "greet" command. You can test the greet command (and maybe other commands) in your `test-deploy` workflow as an integration test. You will be able to execute the commands to validate they run without error, and could even verify their functionality by running additional checks.
```

> You should see a job by default in your `test-deploy.yml` file named `command-tests`. This example job will run one or all of our commands as an integration tests.

Reviewer 2: ```suggestion
You should see a job in your `test-deploy.yml` file named `command-tests`. This example job will run one or all of your commands as an integration test.
```

> In it, we will call our orb command, with any parameters we want to test. If our command were to do something such as, installing a command line tool, we can then test to ensure that command is valid in an additional step.

Reviewer 2: ```suggestion
In this job, you can call your orb command, with any parameters you want to test. If your command, for example, installs a command line tool, you can test to ensure that command is valid in an additional step.
```

> By default you will see the included "greet" command is being tested. Because the greet command only outputs a message to stdout, we can not do any additional validation checks.

Reviewer 2: ```suggestion
By default you will see the included "greet" command is being tested. Because the greet command only outputs a message to stdout, you can not do any additional validation checks.
```

```yaml
jobs:
    command-tests:
      docker:
        - image: cimg/base:current
      steps:
        # Run your orb's commands to validate them.
        - <orb-name>/greet
```

Here is a snippet of a real example from our [GitHub CLI orb](https://github.com/CircleCI-Public/github-cli-orb):

```yaml
jobs:
    command-tests:
      docker:
        - image: cimg/base:current
      steps:
        - github-cli/install
        - run:
            name: verify Install
            command: command -v gh
```

> Here, we are testing the `github-cli/install` command. This command may pass or fail on its own, but we can also validate in the next step that the GitHub CLI has been installed and is available on the command line. If the `gh` binary is not found in the path, this job will fail at this step.

Reviewer 2: ```suggestion
In this example we are testing the `github-cli/install` command. This command may pass or fail on its own, but we can also validate in the next step that the GitHub CLI has been installed and is available on the command line. If the `gh` binary is not found in the path, this job will fail at this step.
```

Remember that you can have multiple jobs for testing commands if desired, or if your orb has no commands, you may have no such job. Just ensure that your `orb-tools/publish` job is requiring any jobs that contain your tests.

### Testing orb jobs
{: #testing-orb-jobs }

> Testing the jobs within our orbs is very similar to testing commands, however there are a few additional restrictions to consider.

Reviewer 2: ```suggestion
Testing jobs within your orbs is very similar to testing commands. However, there are a few additional restrictions to consider.
```

> First, in your `test-deploy` workflow, you will see, just as we mentioned with testing commands above, there is ultimately an `orb-tools/publish` job which requires every job before it in the workflow. To test the jobs of our orb, we simply need to add them to this workflow and ensure they are required in the `orb-tools/publish` job.

Reviewer 2: ```suggestion
First, in your `test-deploy` workflow, you will see, just as we mentioned with testing commands above, there is ultimately an `orb-tools/publish` job which requires every job before it in the workflow to have completed. To test the jobs of your orb, you need to add them to this workflow and ensure they are required in the `orb-tools/publish` job.
```

> Here is an example from our [AWS ECR orb](https://github.com/CircleCI-Public/aws-ecr-orb/blob/0c27bfab932b60f1c60a4c2e74bee114f8d4b795/.circleci/test-deploy.yml#L40)

Reviewer 2: ```suggestion
Here is an example from CircleCI's [AWS ECR orb](https://github.com/CircleCI-Public/aws-ecr-orb/blob/0c27bfab932b60f1c60a4c2e74bee114f8d4b795/.circleci/test-deploy.yml#L40)
```

### Review
{: #review }

The orb-tools orb includes a job `orb-tools/review` which will run a suite of tests against your orb designed to find opportunities to implement best practices and improve the quality of the orb. The "review" job was modeled closely after _ShellCheck_, and operates based on a list of rules called "RC" Review Checks. Each "RC" code corresponds to a specific rule, which can optionally be ignored.

Review Checks output to JUNIT XML formatted and are automatically uploaded to CircleCI to be displayed natively in the UI.

![orb-tools review check RC008]({{site.baseurl}}/assets/img/docs/orbtools-rc008.png)

When you click into the error you will receive more information such as what file and at what line in the code the error was found, along with suggestions for resolution.

#### Local Review

The "review" job is built using the [BATS-Core](https://github.com/bats-core/bats-core) bash automation testing framework. If you have bats installed locally, you can review your orb by running:

```shell
bats src/scripts/review.bats
```

Or, using CircleCI's Local Execute:
```shell
circleci local execute --job orb-tools/review
```

## Unit testing
{: #unit-testing }

> If you are taking advantage of the orb development kit's [`<<include(file)>>` file inclusion]({{site.baseurl}}/2.0/orb-concepts/#file-include-syntax) feature and `src/scripts` directory to store and source your bash files, you can write true integration tests for your scripts.

Reviewer 3: Have "orb development kit" with upper-case initials to keep it consistent with the rest of the docs.
